File: unity/Rotgrid/Assets/Scripts/World/Barricade.cs
using System.Collections.Generic;
using UnityEngine;

namespace Rotgrid
{
    /// <summary>
    /// A boarded window. Zombies queue outside, tear planks off one at a time,
    /// then vault the sill. Players can renail planks for points.
    /// </summary>
    public class Barricade
    {
        public const int BoardsPerWindow = 6;

        public readonly string id;
        public readonly string zone;
        public readonly float x, y, z;
        public readonly float nx, nz;
        public readonly Vector3 outside;   // where attackers stage
        public readonly Vector3 inside;    // where a vault lands
        public readonly Vector3 alcoveCenter;
        public readonly HashSet<object> attackers = new HashSet<object>();

        readonly GameObject[] _boards = new GameObject[BoardsPerWindow];
        readonly bool[] _intact = new bool[BoardsPerWindow];
        readonly float[] _hp = new float[BoardsPerWindow];
        readonly Vector3[] _home = new Vector3[BoardsPerWindow];
        public readonly GameObject root;
        readonly float _wallHeight;

        public Barricade(WindowDef def, float floorY, Transform parent, float wallHeight = 5.2f)
        {
            _wallHeight = wallHeight;
            id = def.id;
            zone = def.zone;
            x = def.x; z = def.z; y = floorY;
            nx = def.nx; nz = def.nz;
            outside = new Vector3(def.x + def.nx * 2.0f, floorY, def.z + def.nz * 2.0f);
            inside = new Vector3(def.x - def.nx * 1.6f, floorY, def.z - def.nz * 1.6f);
            alcoveCenter = new Vector3((def.ax0 + def.ax1) * 0.5f, floorY, (def.az0 + def.az1) * 0.5f);

            root = new GameObject("window_" + def.id);
            root.transform.SetParent(parent, false);
            root.transform.position = new Vector3(def.x, floorY, def.z);
            root.transform.rotation = Quaternion.Euler(0f, Mathf.Atan2(def.nx, def.nz) * Mathf.Rad2Deg, 0f);

            BuildFrame();
            for (int i = 0; i < BoardsPerWindow; i++) MakeBoard(i);
        }

        public int BoardCount
        {
            get
            {
                int n = 0;
                for (int i = 0; i < BoardsPerWindow; i++) if (_intact[i]) n++;
                return n;
            }
        }

        public bool IsOpen { get { return BoardCount == 0; } }

        void BuildFrame()
        {
            var m = Props.Concrete;
            // The wall builder cuts a full-height gap for every window, so the
            // frame fills everything except the aperture. Miss the span above
            // the head and you get a black hole punched through the wall.
            const float ApertureTop = 2.6f;
            const float HeadH = 0.9f;
            Props.Box(3.2f, 1.05f, 0.5f, m, new Vector3(0, 0.52f, 0), root.transform);
            Props.Box(3.2f, HeadH, 0.5f, m, new Vector3(0, ApertureTop + HeadH * 0.5f, 0), root.transform);
            Props.Box(0.4f, 1.6f, 0.5f, m, new Vector3(-1.6f, 1.85f, 0), root.transform);
            Props.Box(0.4f, 1.6f, 0.5f, m, new Vector3(1.6f, 1.85f, 0), root.transform);

            float infillBase = ApertureTop + HeadH;
            float infillH = Mathf.Max(0.1f, _wallHeight + 0.4f - infillBase);
            Props.Box(3.24f, infillH, 0.5f, m, new Vector3(0, infillBase + infillH * 0.5f, 0), root.transform);

            // The aperture stays open: the alcove behind it is a sealed shell, so
            // you can watch zombies crowd the gap and work the planks loose before
            // one gets through. A dim unlit panel deep in the alcove backlights
            // them into silhouettes, far cheaper than a light per window.
            // A Unity quad's normal points down -z, which is already back toward
            // the window, so it needs no rotation.
            Props.Quad(5.2f, 3.4f, Art.Unlit("m_alcoveHaze", Util.Hex(0x46586c)),
                new Vector3(0, 1.9f, 6.15f), root.transform);
        }

        void MakeBoard(int i)
        {
            float by = 1.28f + (i % 3) * 0.5f + (i > 2 ? 0.06f : 0f);
            float tilt = ((i * 37) % 13 - 6) * 1.26f;      // degrees
            float off = ((i * 53) % 11 - 5) * 0.045f;
            float depth = i > 2 ? -0.16f : 0.16f;

            var plank = new GameObject("plank" + i);
            plank.transform.SetParent(root.transform, false);
            plank.transform.localPosition = new Vector3(off, by, depth);
            plank.transform.localRotation = Quaternion.Euler(0, 0, tilt);
            Props.Box(2.7f, 0.19f, 0.09f, Props.Plank, Vector3.zero, plank.transform);
            Props.Box(0.03f, 0.03f, 0.13f, Props.Metal, new Vector3(-1.19f, 0, 0), plank.transform);
            Props.Box(0.03f, 0.03f, 0.13f, Props.Metal, new Vector3(1.19f, 0, 0), plank.transform);

            _boards[i] = plank;
            _home[i] = plank.transform.localPosition;
            _intact[i] = true;
            _hp[i] = 1f;
        }

        /// <summary>A zombie tears at the boards. Returns true when one comes off.</summary>
        public bool Damage(float amount, Effects fx)
        {
            int last = -1;
            for (int i = 0; i < BoardsPerWindow; i++) if (_intact[i]) last = i;
            if (last < 0) return false;

            _hp[last] -= amount;
            // Shake the plank while it is being worked on.
            var t = _boards[last].transform;
            t.localPosition = _home[last] + new Vector3(Util.Rand(-0.02f, 0.02f), 0, 0);

            if (_hp[last] <= 0f)
            {
                _intact[last] = false;
                _boards[last].SetActive(false);
                if (fx != null) fx.WoodBurst(t.position);
                Game.Audio.PlayAt(Sfx.Board, new Vector3(x, y + 1.5f, z), 1f);
                return true;
            }
            return false;
        }

        /// <summary>Player renails one plank. Returns the points awarded.</summary>
        public int Repair(Effects fx)
        {
            for (int i = 0; i < BoardsPerWindow; i++)
            {
                if (_intact[i]) continue;
                _intact[i] = true;
                _hp[i] = 1f;
                _boards[i].SetActive(true);
                _boards[i].transform.localPosition = _home[i];
                Game.Audio.PlayAt(Sfx.Repair, new Vector3(x, y + 1.5f, z), 1f);
                if (fx != null) fx.SparkBurst(_boards[i].transform.position, Util.Hex(0xd9b46a), 6);
                return 10;
            }
            return 0;
        }

        public void RestoreAll()
        {
            for (int i = 0; i < BoardsPerWindow; i++)
            {
                if (_intact[i]) continue;
                _intact[i] = true;
                _hp[i] = 1f;
                _boards[i].SetActive(true);
                _boards[i].transform.localPosition = _home[i];
            }
        }

        /// <summary>True when a point is on the interior side of the wall.</summary>
        public bool IsInside(float px, float pz)
        {
            return (px - x) * nx + (pz - z) * nz < 0f;
        }

        public float DistanceTo(float px, float pz)
        {
            float dx = px - x, dz = pz - z;
            return Mathf.Sqrt(dx * dx + dz * dz);
        }
    }
}

File: src/world/Barricade.js
import * as THREE from 'three';
import { box, Mats } from './Props.js';
import { Audio } from '../core/AudioEngine.js';
import { rand, clamp } from '../core/Util.js';

export const BOARDS_PER_WINDOW = 6;
const BOARD_W = 2.7;

/**
 * A boarded window. Zombies queue at the outside, tear planks off one at a
 * time, then vault the sill. Players can renail planks for points.
 */
export class Barricade {
  constructor(def, floorY, wallHeight = 5.2) {
    this.id = def.id;
    this.wallHeight = wallHeight;
    this.zone = def.zone;
    this.x = def.x;
    this.z = def.z;
    this.y = floorY;
    this.nx = def.nx;
    this.nz = def.nz;
    this.alcove = def.alcove;

    // Outside staging point and the interior landing spot.
    this.outside = { x: def.x + def.nx * 2.0, z: def.z + def.nz * 2.0 };
    this.inside = { x: def.x - def.nx * 1.6, z: def.z - def.nz * 1.6 };

    this.boards = [];
    this.attackers = new Set();
    this.repairCooldown = 0;
    this.group = new THREE.Group();
    this.group.position.set(def.x, floorY, def.z);
    this.group.rotation.y = Math.atan2(def.nx, def.nz);

    this._buildFrame();
    for (let i = 0; i < BOARDS_PER_WINDOW; i++) this._makeBoard(i);
  }

  get boardCount() { return this.boards.filter((b) => b.intact).length; }
  get isOpen() { return this.boardCount === 0; }

  _buildFrame() {
    const frameMat = Mats.concrete;
    // The wall builder cuts a full-height gap for every window, so the frame
    // is responsible for filling everything except the aperture itself. Miss
    // the span above the head and you get a black hole punched through the
    // wall — which is exactly what it looks like from inside.
    const APERTURE_TOP = 2.6;
    const HEAD_H = 0.9;
    this.group.add(box(3.2, 1.05, 0.5, frameMat, 0, 0.52, 0));
    this.group.add(box(3.2, HEAD_H, 0.5, frameMat, 0, APERTURE_TOP + HEAD_H / 2, 0));
    this.group.add(box(0.4, 1.6, 0.5, frameMat, -1.6, 1.85, 0));
    this.group.add(box(0.4, 1.6, 0.5, frameMat, 1.6, 1.85, 0));

    // Infill from the head up to the ceiling. Slightly oversized so it never
    // leaves a hairline seam against the surrounding wall.
    const infillBase = APERTURE_TOP + HEAD_H;
    const infillH = Math.max(0.1, this.wallHeight + 0.4 - infillBase);
    const infill = box(3.24, infillH, 0.5, frameMat, 0, infillBase + infillH / 2, 0);
    infill.castShadow = false;
    this.group.add(infill);

    // The aperture stays open: the alcove behind it is a sealed shell, so you
    // can watch zombies crowd the gap and work the planks loose before the
    // first one gets through. A dim panel deep in the alcove backlights them
    // into silhouettes — far cheaper than a real light per window, and it
    // reads better besides.
    const haze = new THREE.Mesh(
      new THREE.PlaneGeometry(5.2, 3.4),
      // Unfogged: the alcove is only a few metres deep but the interior fog is
      // thick enough to swallow the panel entirely if it participates.
      new THREE.MeshBasicMaterial({ color: 0x46586c, fog: false }),
    );
    haze.position.set(0, 1.9, 6.15);
    haze.rotation.y = Math.PI;
    this.group.add(haze);
  }

  _makeBoard(i) {
    const y = 1.28 + (i % 3) * 0.5 + (i > 2 ? 0.06 : 0);
    const tilt = ((i * 37) % 13 - 6) * 0.022;
    const off = ((i * 53) % 11 - 5) * 0.045;
    const m = box(BOARD_W, 0.19, 0.09, Mats.plank, off, y, i > 2 ? -0.16 : 0.16);
    m.rotation.z = tilt;
    m.userData.baseY = y;
    this.group.add(m);
    // Nails.
    for (const s of [-1, 1]) {
      const n = box(0.03, 0.03, 0.13, Mats.metal, off + s * (BOARD_W / 2 - 0.16), y, (i > 2 ? -0.16 : 0.16));
      n.rotation.z = tilt;
      this.group.add(n);
      m.userData.nails = (m.userData.nails || []).concat(n);
    }
    const b = { mesh: m, intact: true, hp: 1, index: i };
    this.boards.push(b);
    return b;
  }

  /** Zombie tears at the barricade. Returns true when a plank comes off. */
  damage(amount, fx) {
    const live = this.boards.filter((b) => b.intact);
    if (!live.length) return false;
    const b = live[live.length - 1];
    b.hp -= amount;
    // Shake the plank while it is being worked on.
    b.mesh.position.x += rand(-0.02, 0.02);
    b.mesh.rotation.z += rand(-0.01, 0.01);
    if (b.hp <= 0) {
      b.intact = false;
      b.mesh.visible = false;
      b.mesh.userData.nails?.forEach((n) => { n.visible = false; });
      const wp = this.group.localToWorld(b.mesh.position.clone());
      fx?.woodBurst(wp);
      Audio.playAt('board', { x: this.x, y: this.y + 1.5, z: this.z }, { volume: 1 });
      return true;
    }
    return false;
  }

  /** Player repairs one plank. Returns points awarded, or 0. */
  repair(fx) {
    const broken = this.boards.find((b) => !b.intact);
    if (!broken) return 0;
    broken.intact = true;
    broken.hp = 1;
    broken.mesh.visible = true;
    broken.mesh.position.x = ((broken.index * 53) % 11 - 5) * 0.045;
    broken.mesh.rotation.z = ((broken.index * 37) % 13 - 6) * 0.022;
    broken.mesh.userData.nails?.forEach((n) => { n.visible = true; });
    Audio.playAt('repair', { x: this.x, y: this.y + 1.5, z: this.z }, { volume: 1 });
    fx?.sparkBurst(this.group.localToWorld(broken.mesh.position.clone()), 0xd9b46a, 6);
    return 10;
  }

  /** Rebuild everything (used by round transitions on empty windows). */
  restoreAll() {
    for (const b of this.boards) {
      if (b.intact) continue;
      b.intact = true;
      b.hp = 1;
      b.mesh.visible = true;
      b.mesh.userData.nails?.forEach((n) => { n.visible = true; });
    }
  }

  distanceTo(x, z) { return Math.hypot(x - this.x, z - this.z); }

  /** True when a player standing at (x,z) is on the interior side. */
  isInside(x, z) {
    return (x - this.x) * this.nx + (z - this.z) * this.nz < 0;
  }

  update(dt) {
    this.repairCooldown = clamp(this.repairCooldown - dt, 0, 10);
  }
}
